service/matching/matchingEngine.go

```diff
@@ -444,10 +446,16 @@ func (e *matchingEngineImpl) DispatchSpooledTask(
 	unversionedOrigTaskQueue := newTaskQueueIDWithVersionSet(origTaskQueue, "")
 	// Redirect and re-resolve if we're blocked in matcher and user data changes.
 	for {
-		taskQueue, userDataChanged, err := e.redirectToVersionedQueueForAdd(
+		taskQueue, _, userDataChanged, err := e.redirectToVersionedQueueForAdd(
 			ctx, unversionedOrigTaskQueue, directive, stickyInfo)
 		if err != nil {
-			return err
+			if errors.Is(err, errUserDataDisabled) && directive.GetBuildId() == "" {
```

Comment:
you could put this `if errors.Is(...) && ... { err = nil }` above the `if err != nil` also